File: x/adrsimon/mobile/Dust/Services/APIClient.swift
import Foundation
import os

private let logger = Logger(subsystem: AppConfig.bundleId, category: "APIClient")

enum APIError: LocalizedError {
    case invalidURL
    case invalidResponse
    case httpError(statusCode: Int, body: String)
    case decodingError(Error)
    case networkError(Error)

    var errorDescription: String? {
        switch self {
        case .invalidURL:
            "Invalid URL"
        case .invalidResponse:
            "Invalid response from server"
        case let .httpError(statusCode, body):
            "HTTP \(statusCode): \(body)"
        case let .decodingError(error):
            "Failed to decode response: \(error.localizedDescription)"
        case let .networkError(error):
            error.localizedDescription
        }
    }
}

enum APIClient {
    private static let snakeCaseDecoder: JSONDecoder = {
        let decoder = JSONDecoder()
        decoder.keyDecodingStrategy = .convertFromSnakeCase
        decoder.dateDecodingStrategy = .iso8601
        return decoder
    }()

    private static let camelCaseDecoder: JSONDecoder = {
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        return decoder
    }()

    private static let encoder: JSONEncoder = {
        let encoder = JSONEncoder()
        encoder.keyEncodingStrategy = .convertToSnakeCase
        return encoder
    }()

    private static let camelCaseEncoder: JSONEncoder = .init()

    // MARK: - Public (raw access token)

    static func get<T: Decodable>(
        _ endpoint: String,
        accessToken: String? = nil,
        snakeCase: Bool = true
    ) async throws -> T {
        var request = try buildRequest(endpoint: endpoint, accessToken: accessToken)
        request.httpMethod = "GET"
        let decoder = snakeCase ? snakeCaseDecoder : camelCaseDecoder
        return try await execute(request, endpoint: endpoint, decoder: decoder)
    }

    static func post<T: Decodable>(
        _ endpoint: String,
        body: some Encodable,
        accessToken: String? = nil
    ) async throws -> T {
        var request = try buildRequest(endpoint: endpoint, accessToken: accessToken)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try encoder.encode(body)
        return try await execute(request, endpoint: endpoint, decoder: snakeCaseDecoder)
    }

    static func postNoResponse(
        _ endpoint: String,
        body: some Encodable,
        accessToken: String? = nil
    ) async throws {
        var request = try buildRequest(endpoint: endpoint, accessToken: accessToken)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try encoder.encode(body)
        _ = try await performRequest(request)
    }

    // MARK: - Public (TokenProvider with automatic 401 retry)

    static func authenticatedGet<T: Decodable>(
        _ endpoint: String,
        tokenProvider: TokenProvider,
        snakeCase: Bool = true
    ) async throws -> T {
        let decoder = snakeCase ? snakeCaseDecoder : camelCaseDecoder
        return try await withAuthRetry(tokenProvider: tokenProvider) { token in
            var request = try buildRequest(endpoint: endpoint, accessToken: token)
            request.httpMethod = "GET"
            return try await execute(request, endpoint: endpoint, decoder: decoder)
        }
    }

    static func authenticatedPost<T: Decodable>(
        _ endpoint: String,
        body: some Encodable,
        tokenProvider: TokenProvider,
        snakeCase: Bool = true
    ) async throws -> T {
        let selectedEncoder = snakeCase ? encoder : camelCaseEncoder
        let selectedDecoder = snakeCase ? snakeCaseDecoder : camelCaseDecoder
        let encodedBody = try selectedEncoder.encode(body)
        return try await withAuthRetry(tokenProvider: tokenProvider) { token in
            var request = try buildRequest(endpoint: endpoint, accessToken: token)
            request.httpMethod = "POST"
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.httpBody = encodedBody
            return try await execute(request, endpoint: endpoint, decoder: selectedDecoder)
        }
    }

    /// Sends an authenticated request with no response body decoding. Retries once on 401.
    static func authenticatedSend(
        _ endpoint: String,
        method: String,
        body: Data,
        tokenProvider: TokenProvider
    ) async throws {
        let _: Bool = try await withAuthRetry(tokenProvider: tokenProvider) { token in
            var request = try buildRequest(endpoint: endpoint, accessToken: token)
            request.httpMethod = method
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.httpBody = body
            _ = try await performRequest(request)
            return true
        }
    }

    /// Uploads file data as multipart/form-data. The URL may be absolute or relative.
    static func authenticatedMultipartUpload<T: Decodable>(
        _ urlString: String,
        fileData: Data,
        fileName: String,
        mimeType: String,
        tokenProvider: TokenProvider,
        snakeCase: Bool = true
    ) async throws -> T {
        let boundary = "Boundary-\(UUID().uuidString)"
        let body = buildMultipartBody(fileData: fileData, fileName: fileName, mimeType: mimeType, boundary: boundary)
        let decoder = snakeCase ? snakeCaseDecoder : camelCaseDecoder

        return try await withAuthRetry(tokenProvider: tokenProvider) { token in
            let fullURL: URL
            if urlString.hasPrefix("http") {
                guard let url = URL(string: urlString) else { throw APIError.invalidURL }
                fullURL = url
            } else {
                guard let url = URL(string: "\(AppConfig.apiBaseURL)\(urlString)") else { throw APIError.invalidURL }
                fullURL = url
            }

            var request = URLRequest(url: fullURL)
            request.httpMethod = "POST"
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
            request.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")
            request.httpBody = body
            return try await execute(request, endpoint: urlString, decoder: decoder)
        }
    }

    private static func buildMultipartBody(
        fileData: Data,
        fileName: String,
        mimeType: String,
        boundary: String
    ) -> Data {
        var body = Data()
        body.append(Data("--\(boundary)\r\n".utf8))
        body.append(Data("Content-Disposition: form-data; name=\"file\"; filename=\"\(fileName)\"\r\n".utf8))
        body.append(Data("Content-Type: \(mimeType)\r\n\r\n".utf8))
        body.append(fileData)
        body.append(Data("\r\n--\(boundary)--\r\n".utf8))
        return body
    }

    /// Executes a closure with a valid access token, retrying once on 401 after refreshing.
    private static func withAuthRetry<T>(
        tokenProvider: TokenProvider,
        _ operation: (String) async throws -> T
    ) async throws -> T {
        let token = try await tokenProvider.validAccessToken()
        do {
            return try await operation(token)
        } catch APIError.httpError(statusCode: 401, _) {
            let freshToken = try await tokenProvider.refreshedAccessToken()
            return try await operation(freshToken)
        }
    }

    private static func buildRequest(endpoint: String, accessToken: String?) throws -> URLRequest {
        guard let url = URL(string: "\(AppConfig.apiBaseURL)\(endpoint)") else {
            throw APIError.invalidURL
        }
        var request = URLRequest(url: url)
        if let accessToken {
            request.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")
        }
        return request
    }

    /// Executes the request and validates the HTTP response, returning the raw data.
    private static func performRequest(_ request: URLRequest) async throws -> Data {
        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await URLSession.shared.data(for: request)
        } catch {
            throw APIError.networkError(error)
        }

        guard let httpResponse = response as? HTTPURLResponse else {
            throw APIError.invalidResponse
        }

        guard (200 ... 299).contains(httpResponse.statusCode) else {
            let body = String(data: data, encoding: .utf8) ?? "Unknown error"
            throw APIError.httpError(statusCode: httpResponse.statusCode, body: body)
        }

        return data
    }

    private static func execute<T: Decodable>(
        _ request: URLRequest,
        endpoint: String,
        decoder: JSONDecoder
    ) async throws -> T {
        let data = try await performRequest(request)

        do {
            return try decoder.decode(T.self, from: data)
        } catch {
            let rawBody = String(data: data, encoding: .utf8) ?? "<non-utf8>"
            logger.error("Decoding failed for \(endpoint). Raw response:\n\(rawBody)")
            logger.error("Decode error: \(error)")
            throw APIError.decodingError(error)
        }
    }
}

File: x/aubin/android/core/src/main/kotlin/com/dust/mobile/core/network/ApiClient.kt
package com.dust.mobile.core.network

import com.dust.mobile.core.auth.TokenProvider
import com.dust.mobile.core.config.AppConfig
import kotlinx.serialization.KSerializer
import kotlinx.serialization.encodeToString
import kotlinx.serialization.serializer
import kotlinx.coroutines.CancellationException
import java.util.UUID

class ApiClient(
    private val config: AppConfig,
    private val engine: HttpEngine,
) {
    suspend inline fun <reified T> get(
        endpoint: String,
        accessToken: String? = null,
    ): T = get(endpoint, serializer(), accessToken)

    suspend fun <T> get(
        endpoint: String,
        serializer: KSerializer<T>,
        accessToken: String? = null,
    ): T {
        val request = buildRequest(HttpMethod.GET, endpoint, accessToken)
        return executeJson(request, serializer)
    }

    suspend inline fun <reified Body, reified T> post(
        endpoint: String,
        body: Body,
        accessToken: String? = null,
    ): T = post(endpoint, body, serializer(), serializer(), accessToken)

    suspend fun <Body, T> post(
        endpoint: String,
        body: Body,
        bodySerializer: KSerializer<Body>,
        responseSerializer: KSerializer<T>,
        accessToken: String? = null,
    ): T {
        val request = buildRequest(
            method = HttpMethod.POST,
            endpoint = endpoint,
            accessToken = accessToken,
            headers = mapOf("Content-Type" to "application/json"),
            body = DustJson.encodeToString(bodySerializer, body).encodeToByteArray(),
        )
        return executeJson(request, responseSerializer)
    }

    suspend inline fun <reified Body> send(
        endpoint: String,
        method: HttpMethod,
        body: Body,
        accessToken: String? = null,
    ) = send(endpoint, method, body, serializer(), accessToken)

    suspend fun <Body> send(
        endpoint: String,
        method: HttpMethod,
        body: Body,
        bodySerializer: KSerializer<Body>,
        accessToken: String? = null,
    ) {
        val request = buildRequest(
            method = method,
            endpoint = endpoint,
            accessToken = accessToken,
            headers = mapOf("Content-Type" to "application/json"),
            body = DustJson.encodeToString(bodySerializer, body).encodeToByteArray(),
        )
        perform(request)
    }

    suspend inline fun <reified T> authenticatedGet(
        endpoint: String,
        tokenProvider: TokenProvider,
    ): T = withAuthRetry(tokenProvider) { token -> get<T>(endpoint, token) }

    suspend inline fun <reified Body, reified T> authenticatedPost(
        endpoint: String,
        body: Body,
        tokenProvider: TokenProvider,
    ): T = withAuthRetry(tokenProvider) { token -> post<Body, T>(endpoint, body, token) }

    suspend inline fun <reified Body> authenticatedSend(
        endpoint: String,
        method: HttpMethod,
        body: Body,
        tokenProvider: TokenProvider,
    ) = withAuthRetry(tokenProvider) { token ->
        send(endpoint, method, body, token)
        Unit
    }

    suspend inline fun <T> withAuthRetry(
        tokenProvider: TokenProvider,
        crossinline operation: suspend (String) -> T,
    ): T {
        val token = tokenProvider.validAccessToken()
        return try {
            operation(token)
        } catch (error: ApiError.Http) {
            if (error.statusCode != 401) {
                throw error
            }
            operation(tokenProvider.refreshedAccessToken())
        }
    }

    suspend fun authenticatedRawGet(
        endpoint: String,
        tokenProvider: TokenProvider,
    ): ByteArray = withAuthRetry(tokenProvider) { token ->
        perform(buildRequest(HttpMethod.GET, endpoint, token)).body
    }

    suspend fun authenticatedRawGetResponse(
        endpoint: String,
        tokenProvider: TokenProvider,
    ): HttpResponse = withAuthRetry(tokenProvider) { token ->
        perform(buildRequest(HttpMethod.GET, endpoint, token))
    }

    suspend inline fun <reified T> authenticatedMultipartUpload(
        urlString: String,
        fileData: ByteArray,
        fileName: String,
        mimeType: String,
        tokenProvider: TokenProvider,
    ): T = authenticatedMultipartUpload(urlString, fileData, fileName, mimeType, tokenProvider, serializer())

    suspend fun <T> authenticatedMultipartUpload(
        urlString: String,
        fileData: ByteArray,
        fileName: String,
        mimeType: String,
        tokenProvider: TokenProvider,
        responseSerializer: KSerializer<T>,
    ): T {
        val boundary = "Boundary-${UUID.randomUUID()}"
        val body = buildMultipartBody(fileData, fileName, mimeType, boundary)
        return withAuthRetry(tokenProvider) { token ->
            val request = buildRequest(
                method = HttpMethod.POST,
                endpoint = urlString,
                accessToken = token,
                headers = mapOf("Content-Type" to "multipart/form-data; boundary=$boundary"),
                body = body,
            )
            executeJson(request, responseSerializer)
        }
    }

    suspend fun authenticatedMultipartRaw(
        endpoint: String,
        fileData: ByteArray,
        fileName: String,
        mimeType: String,
        tokenProvider: TokenProvider,
    ): ByteArray {
        val boundary = "Boundary-${UUID.randomUUID()}"
        val body = buildMultipartBody(fileData, fileName, mimeType, boundary)
        return withAuthRetry(tokenProvider) { token ->
            val request = buildRequest(
                method = HttpMethod.POST,
                endpoint = endpoint,
                accessToken = token,
                headers = mapOf("Content-Type" to "multipart/form-data; boundary=$boundary"),
                body = body,
            )
            perform(request).body
        }
    }

    fun buildMultipartBody(
        fileData: ByteArray,
        fileName: String,
        mimeType: String,
        boundary: String,
    ): ByteArray {
        val prefix = "--$boundary\r\n" +
            "Content-Disposition: form-data; name=\"file\"; filename=\"$fileName\"\r\n" +
            "Content-Type: $mimeType\r\n\r\n"
        val suffix = "\r\n--$boundary--\r\n"
        return prefix.encodeToByteArray() + fileData + suffix.encodeToByteArray()
    }

    suspend fun perform(request: HttpRequest): HttpResponse {
        val response = try {
            engine.execute(request)
        } catch (error: CancellationException) {
            throw error
        } catch (error: Throwable) {
            throw ApiError.Network(error)
        }
        if (response.statusCode !in 200..299) {
            throw ApiError.Http(response.statusCode, response.body.decodeToString())
        }
        return response
    }

    fun buildRequest(
        method: HttpMethod,
        endpoint: String,
        accessToken: String? = null,
        headers: Map<String, String> = emptyMap(),
        body: ByteArray? = null,
    ): HttpRequest {
        val url = if (endpoint.startsWith("http://") || endpoint.startsWith("https://")) {
            endpoint
        } else {
            "${config.apiBaseUrl}$endpoint"
        }
        val authHeaders = if (accessToken == null) {
            emptyMap()
        } else {
            mapOf("Authorization" to "Bearer $accessToken")
        }
        return HttpRequest(method = method, url = url, headers = authHeaders + headers, body = body)
    }

    private suspend fun <T> executeJson(request: HttpRequest, serializer: KSerializer<T>): T {
        val response = perform(request)
        val text = response.body.decodeToString()
        return try {
            DustJson.decodeFromString(serializer, text)
        } catch (error: CancellationException) {
            throw error
        } catch (error: Throwable) {
            throw ApiError.Decoding(text, error)
        }
    }
}
